### PR TITLE
[Model] Support post-norm architecture for EAGLE-3 supeculators

### PR DESCRIPTION
**Pull request by Dogacel**
## Purpose

Support the post-norm architecture EAGLE-3 models for speculative decoding.

Blog / Twitter Thread: https://x.com/dogacel0/status/2054203929378873661?s=20

1. Support post-norm & individual fc norm architectural changes.
2. Fix a bug in `gpu_model_runner.py` where captured aux layers were not properly configured.

## Test Plan

Run the following script for existing and new gpt-oss speculators.

Existing:
```
num_spec_tokens     : 5
batch (max_num_seqs): 8
num_drafts          : 6590
mean acceptance len : 2.359
  accept@pos 0      : 0.622  (4098)
  accept@pos 1      : 0.356  (2344)
  accept@pos 2      : 0.203  (1339)
  accept@pos 3      : 0.119  (783)
  accept@pos 4      : 0.060  (394)
```

Ours:
```
num_spec_tokens     : 5
batch (max_num_seqs): 8
num_drafts          : 7431
mean acceptance len : 2.346
  accept@pos 0      : 0.610  (4530)
  accept@pos 1      : 0.342  (2544)
  accept@pos 2      : 0.202  (1500)
  accept@pos 3      : 0.120  (891)
  accept@pos 4      : 0.072  (538)
```

The acceptance rate increases especially on deeper positions, longer context and template changes. Those tests just show that both existing and old models work as expected.

```python
# SPDX-License-Identifier: Apache-2.0
# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
# spec_decode_gptoss.py
import argparse

from datasets import load_dataset
from transformers import AutoTokenizer

from vllm import LLM, SamplingParams
from vllm.v1.metrics.reader import Counter, Vector

TARGET = "openai/gpt-oss-120b"
# DRAFT  = "RedHatAI/gpt-oss-120b-speculator.eagle3"
# DRAFT = "nvidia/gpt-oss-120b-Eagle3-v3"
DRAFT = "Dogacel/specdrift-gpt-oss-120b-eagle3"

# TARGET = "openai/gpt-oss-20b"
# DRAFT  = "RedHatAI/gpt-oss-20b-speculator.eagle3"
# DRAFT = "Dogacel/specdrift-gpt-oss-20b-eagle3"


def main():
    p = argparse.ArgumentParser()
    p.add_argument("--num_spec_tokens", type=int, default=5)
    p.add_argument("--num_prompts", type=int, default=20)
    p.add_argument("--max_num_seqs", type=int, default=8)
    p.add_argument(
        "--tp", type=int, default=1
    )  # 120B fits on 1xH100 (MXFP4); bump to 2/4 for KV headroom
    p.add_argument("--max_tokens", type=int, default=2048)
    args = p.parse_args()

    tok = AutoTokenizer.from_pretrained(TARGET)
    ds = load_dataset("philschmid/mt-bench", split="train")
    prompts = [
        tok.apply_chat_template(
            [{"role": "user", "content": ex["turns"][0]}],
            tokenize=False,
            add_generation_prompt=True,
        )
        for ex in ds.select(range(args.num_prompts))
    ]

    llm = LLM(
        model=TARGET,
        tensor_parallel_size=1,
        max_num_seqs=8,  # ← drop way down
        max_model_len=4096,  # ← drop further
        gpu_memory_utilization=0.95,  # ← push higher
        enforce_eager=True,  # ← per the RedHatAI 120b card
        enable_prefix_caching=False,
        disable_log_stats=False,
        speculative_config={
            "method": "eagle3",
            "model": DRAFT,
            "num_speculative_tokens": 5,
        },
    )

    sp = SamplingParams(temperature=0.7, max_tokens=args.max_tokens)
    llm.generate(prompts, sp)

    # Pull V1 spec-decode metrics
    metrics = llm.get_metrics()
    num_drafts = num_accepted = 0
    per_pos = None
    for m in metrics:
        if isinstance(m, Counter) and m.name == "vllm:spec_decode_num_drafts":
            num_drafts = m.value
        elif (
            isinstance(m, Counter) and m.name == "vllm:spec_decode_num_accepted_tokens"
        ):
            num_accepted = m.value
        elif (
            isinstance(m, Vector)
            and m.name == "vllm:spec_decode_num_accepted_tokens_per_pos"
        ):
            per_pos = m.values

    mean_al = 1 + num_accepted / max(num_drafts, 1)
    print("-" * 60)
    print(f"num_spec_tokens     : {args.num_spec_tokens}")
    print(f"batch (max_num_seqs): {args.max_num_seqs}")
    print(f"num_drafts          : {num_drafts}")
    print(f"mean acceptance len : {mean_al:.3f}")
    if per_pos:
        for i, c in enumerate(per_pos):
            rate = c / max(num_drafts, 1)
            print(f"  accept@pos {i}      : {rate:.3f}  ({c})")


if __name__ == "__main__":
    main()
```

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/v1/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -5167,9 +5167,14 @@ def _get_eagle3_aux_layers_from_config(self) -> tuple[int, ...] | None:
         layer_ids = getattr(hf_config, "eagle_aux_hidden_state_layer_ids", None)
         if not layer_ids:
             dflash_config = getattr(hf_config, "dflash_config", None)
+            eagle_config = getattr(hf_config, "eagle_config", None)
+
             if dflash_config and isinstance(dflash_config, dict):
                 layer_ids = dflash_config.get("target_layer_ids")
 
+            if eagle_config and isinstance(eagle_config, dict):
+                layer_ids = eagle_config.get("eagle_aux_hidden_state_layer_ids")
```
[the remark this answers is not in the document]
Comment (Dogacel) [reply vllm/v1/worker/gpu_model_runner.py:5176]: D-Flash and EAGLE are independent architectures, this should never happen.

File: vllm/model_executor/models/llama_eagle3.py [hunk 2]
```diff
@@ -371,6 +402,14 @@ def combine_hidden_states(
 
         if self.model.norm_before_fc:
             hidden_states = self.model.input_norm(hidden_states)
+
+        if self.model.fc_norm is not None:
+            chunks = hidden_states.chunk(self.model.num_aux_hidden_states, dim=-1)
+            hidden_states = torch.cat(
+                [norm(chunk) for norm, chunk in zip(self.model.fc_norm, chunks)],
+                dim=-1,
+            )
```
Comment (TheEpicDolphin) [inline vllm/model_executor/models/llama_eagle3.py:405]: It could be a bit confusing to future readers to see two branches here applying post-norm in different ways. I think a comment here explaining how they differ would be helpful. Maybe we can emphasize that only one branch or the other is ever active at a time?

File: vllm/model_executor/models/llama_eagle3.py [hunk 3]
```diff
@@ -310,13 +337,17 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
...
+                self.config, "target_hidden_size", self.config.hidden_size
+            )
             self.register_buffer(
                 "mask_hidden",
-                torch.zeros(
-                    1,
-                    (3 if self.model.use_aux_hidden_state else 1)
-                    * self.config.hidden_size,
-                ),
+                torch.zeros(1, aux_count * target_hidden_size),
```
Comment (TheEpicDolphin) [inline vllm/model_executor/models/llama_eagle3.py]: nit: If you make `fc_input_size` from above into a property of `LlamaModel`, you can use it here instead of recomputing

Comment (Dogacel) [reply vllm/model_executor/models/llama_eagle3.py]: Done 👍

File: vllm/model_executor/models/llama_eagle3.py (hunk 2 above)
Comment (Dogacel) [reply vllm/model_executor/models/llama_eagle3.py:405]: Added 👍

Comment (Dogacel) [reply vllm/model_executor/models/llama_eagle3.py:405]: I've basically explained they are two different ways of applying normalization.

**Review by TheEpicDolphin — APPROVED** [review]
Thanks for adding this feature! These changes look good to me. The failing model-executor test seems unrelated to this PR (recently fixed here: https://github.com/vllm-project/vllm/pull/42783).

cc: @simon-mo, @WoosukKwon

**Review by simon-mo — APPROVED** [review]
